CSR siselect/sireg: Supervisor indirect register select/alias

ted-logan opened the pull request:
Implement external interrupt CSRs siselect/sireg:

- siselect: Supervisor indirect register select
- sireg: Supervisor indirect register alias

These CSRs are specified in the RISC-V Advanced Interrupt Architecture (AIA) (https://github.com/riscv/riscv-aia/releases).

furquan-rivos commented on src/register/sireg.rs:
```diff
@@ -0,0 +1,162 @@
+//! Architecture" Version 0.3.2-draft
+//!
+//! Before accessing `sireg` using the functions in this module,
+//! software must select the indirect register using the `siselect` CSR.
```
Why give the flexibility to software to make mistakes? I understand rest of this crate is organized such that it gives individual CSR access to caller. But, siselect/sireg are special. These are window registers that provide access to a large set of registers. What do you think about only providing access to caller to perform:

```
sireg_read_eidelivery() -> usize
sireg_write_eidelivery(usize)

sireg_read_eithreshold() -> usize
sireg_write_eithreshold(usize)

sireg_read_eip(usize) -> EipStatus
sireg_write_eip(usize, EipStatus)

sireg_read_eie(usize) -> EieStatus
sireg_write_eie(usize, EieStatus)
```

For eip and eie we don't even need to expose the actual reg # because the caller will have to calculate which bit in which eipX / eieX register corresponds to an interrupt #. Thus, instead these functions can take interrupt # as input and do the necessary calculation for determining what CSR to read/write.

ted-logan replied:
Yeah, I think you're right. I was trying to find the right balance in the API between making the code simple enough and matching the style of the rest of the crate; versus making the siselect/sireg interface easier (and safer) to use. I've rewritten this interface to make sireg the only interface one needs to use to access the registers (and improved the interrupt enables and interrupt pending bits by making the sireg interface aware of how the bits are packed into registers, to make the interface easier to use).

furquan-rivos replied:
As I was reading your code, I realized that we could have solved this in another way, probably slightly more cleaner w.r.t. internal implementation and from caller standpoint.

1. Make sireg and siselect both private i.e. callers never see the regs directly. These are only for internal usage.
2. Create public facing CSRs for eidelivery, eithreshold, eip, eie and provide implementations that make use of sireg and siselect to do their work. We can also provide some riscv-specific register macros to provide the boilerplate code for `read_windowed_csr`/`write_windowed_csr`.

Anyways, I think what you have here is easy to use as well. From a caller standpoint, the only difference would have been: Instead of using `register::sireg_eidelivery::read()`, they would use `register::eidelivery::read()`.

When we move to tock registers, maybe we can keep that in mind.

furquan-rivos replied:
As I was reading your code, I realized that we could have solved this in another way, probably slightly more cleaner w.r.t. internal implementation and from caller standpoint.

Make sireg and siselect both private i.e. callers never see the regs directly. These are only for internal usage.
Create public facing CSRs for eidelivery, eithreshold, eip, eie and provide implementations that make use of sireg and siselect to do their work. We can also provide some riscv-specific register macros to provide the boilerplate code for read_windowed_csr/write_windowed_csr.
Anyways, I think what you have here is easy to use as well. From a caller standpoint, the only difference would have been: Instead of using register::sireg_eidelivery::read(), they would use register::eidelivery::read().

When we move to tock registers, maybe we can keep that in mind.

ted-logan committed:
CSR siselect/sireg: Supervisor indirect register select/alias

Implement external interrupt CSRs siselect/sireg:

- siselect: Supervisor indirect register select
- sireg: Supervisor indirect register alias

These CSRs are specified in the RISC-V Advanced Interrupt Architecture
(AIA) (https://github.com/riscv/riscv-aia/releases).

furquan-rivos commented on src/register/mod.rs:
```diff
@@ -62,6 +62,8 @@ pub mod stimecmp;
 
 // Supervisor-Level Interrupts
+pub mod sireg;
+pub mod siselect;
```
Since the caller is not expected to deal with siselect directly, does it make sense to not make siselect module public and only include it within sireg?

adurbin-rivos replied:
I would argue one shouldn't be accessing things directly through sireg language in that case either since it's one piece of an index/data pair access. It seems more appropriate to wrap in an aia library dealing w/ those particular nuances.

furquan-rivos replied:
Yeah, I posted a comment along those lines [here](https://github.com/rivosinc/riscv/pull/10#discussion_r1014529581). I think we should make both `sireg` and `siselect` as internal implementation details. The caller should be exposed the windowed registers(`eie`, `eip`, etc.) and their implementation should deal with using the internal `sireg` and `siselect`.

furquan-rivos commented on src/register/siselect.rs:
```diff
@@ -0,0 +1,156 @@
+pub enum Register {
+    Eidelivery = 0x70,
+    Eithreshold = 0x72,
+    Eip0 = 0x80,
```
nit: You already have the enums all instantiated, so its okay to keep it this way. But, with the recent update, you don't really need to define all Eip* and Eie* register addresses. There can be a single EipBase, EieBase and then calculate the required address from that.
